File: rec-initializer/src/main/java/net/kimleo/rec/init/Initializer.java
package net.kimleo.rec.init;

import java.io.*;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import static net.kimleo.rec.util.Sys.die;

public class Initializer {
    final String file;
    final HashMap<String, String> parameters;
    static final List<String> REQUIRED_PROPERTIES = Arrays.asList("name", "delimiter", "escape", "format");

    public Initializer(String file, HashMap<String, String> parameters) {
        this.file = file;
        this.parameters = parameters;
    }

    public void init() throws IOException {
        String recFile = file + ".rec";

        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        for (String property : REQUIRED_PROPERTIES) {
            if (parameters.containsKey(property)) {
                System.out.printf("Please input the %s:", property);
                String value = reader.readLine().trim();
                if (!checkProperty(property, value)) {
                    die(String.format("Unexpected %s: %s", property, value));
                }
                parameters.put(property, value);
            }
        }

        try (PrintWriter writer =
                     new PrintWriter(new OutputStreamWriter(new FileOutputStream(new File(recFile))))) {
            writer.println(parameters.get("name"));

            for (String property : parameters.keySet()) {
                if (!"name".equals(property)) {
                    writer.println(String.format("%s=%s", property, parameters.get(property)));
                }
            }
        } catch (Exception e) {
            die("Unexpected exception found: " + e.getMessage());
        }
    }

    private boolean checkProperty(String property, String value) {
        if (value == null || value.isEmpty()) return false;
        switch (property) {
            case "name":
                return value.chars().allMatch(Character::isJavaIdentifierPart);
            case "delimiter":
            case "escape":
                return value.length() == 1 || value.matches("^%\\d+&");
        }
        return true;
    }
}

File: rec-initializer/src/main/kotlin/net/kimleo/rec/init/Initializer.kt
package net.kimleo.rec.init

import net.kimleo.rec.exit
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.util.*

class Initializer(val file: String, val parameters: HashMap<String, String> = linkedMapOf()) {

    val requiredProperties = listOf("name", "delimiter", "escape", "format")

    fun init() {
        val recFile = file + ".rec"

        val reader = BufferedReader(InputStreamReader(System.`in`));
        for (property in requiredProperties) {
            if (property !in parameters) {
                println("Please input the $property: ")
                val value = reader.readLine()
                if (!checkProperty(property, value)) { exit("Unexpected $property: $value") }
                parameters[property] = value
            }
        }

        val outputFile = File(recFile)
        val writer = outputFile.writer()
        writer.appendln(parameters["name"])

        for (property in parameters.keys) {
            if (property != "name") {
                writer.appendln("$property=${parameters[property]}")
            }
        }

        writer.close()
    }

    private fun checkProperty(property: String, value: String): Boolean {
        when (property) {
            "name" -> return value.all(Char::isJavaIdentifierPart)
            "delimiter", "escape" -> return value.length == 1 || (value.matches(Regex("^%\\d+$")))
        }
        return true
    }
}